assets/js/src/frontend.js

```diff
@@ -112,112 +48,102 @@ const fetchPosts = ( url, container, clickedButton ) => {
 			throw new Error( 'Network response was not ok.' );
 		} )
 		.then( function ( data ) {
-			// Get the posts container.
+			// create temporary container to load fetched HTML
 			const temp = document.createElement( 'div' );
 			temp.innerHTML = data;
-			const posts = temp.querySelector( '.wp-block-post-template' );
 
-			// Update the posts container.
-			container.insertAdjacentHTML( 'beforeend', posts.innerHTML );
+			// get region from container
+			const containerRegion = container.dataset.qllmQueryRegion || '';
 
-			// Update the window URL.
-			window.history.pushState( {}, '', url );
+			// find container in fetched HTML matching container's region
+			const posts = temp.querySelector(
+				`.wp-block-post-template[data-qllm-query-region="${ containerRegion }"]`
+			);
 
-			const $button = clickedButton.closest('.wp-block-button');
+			// append the posts
+			if ( posts ) {
+				container.insertAdjacentHTML( 'beforeend', posts.innerHTML );
+			}
 
-			console.log( $button );
+			const $button = button.closest( '.wp-block-button' );
 
 			if ( $button ) {
 				$button.classList.remove( 'loading' );
 			}
-
-			// Remove button.
-			clickedButton.remove();
-
-			hideLoader();
 		} )
-		.catch( function ( error ) {
+		.catch( ( error ) => {
+			//eslint-disable-next-line no-console
 			console.error( 'Fetch error:', error );
-		} );
-};
+		} )
+		//cleanup
+		.finally( () => {
+			//no more posts available -> remove button
+			if (
+				+button.dataset.queryNextPage >= +button.dataset.queryMaxPage
```

Comment:
TIL:
 - Casting to integer with `+` instead of `parseInt()`
 - The `dataset` attr to access all pieces at once. 

Very nice!